NEXT-23394 - Update documentation Flow Builder for next major (#703)
shopware/docs: guides/plugins/plugins/framework/flow/add-flow-builder-action.md

interface TagAware extends FlowEventAware
{
    ...

    public const TAG = 'tag';

    public const TAG_ID = 'tagId';

    public function getTag();

    ...
}
```

{% endcode %}

### Create new action

{% hint style="warning" %}

> Since the next major version(`v6.5.0.0`), we have made some changes in how to create new flow actions, please follow the instructions corresponding to your current version.

Reviewer: ```suggestion
Starting with the next major version (6.5.0.0), we are introducing changes in how to create new flow actions, please follow the instructions corresponding to your current version.
```

{% endhint %}

In this example, we will name it `CreateTagAction`. It will be placed in the directory `<plugin root>/src/Core/Content/Flow/Dispatching/Action`. Below you can find an example implementation:

{% hint style="info" %}
  Available starting with Shopware 6.4.6.0
{% endhint %}

> Our new class has to extend from the abstract class `Shopware\Core\Framework\Event\FLowEvent`.

Reviewer: ```suggestion

Our new class has to extend from the abstract class `Shopware\Core\Framework\Event\FLowEvent`.
```

- `requirements`: This defines which interfaces that the action belongs to.
- `handle`: Use this method to handle your action stuff.

{% hint style="info" %}
  Available starting with Shopware 6.5.0.0
{% endhint %}

> {% code title="<plugin root>/src/Core/Content/Flow/Dispatching/Action/CreateTagAction.php" %}

Reviewer: ```suggestion

{% code title="<plugin root>/src/Core/Content/Flow/Dispatching/Action/CreateTagAction.php" %}
```

    {
        // config is the config data when created a flow sequence
        $config = $flow->getConfig();

        // make sure your tags data is exist
        if (!\array_key_exists('tags', $config)) {
            return;
        }

        $tags = $config['tags'];

        // just a step to make sure you're dispatching correct action
        if (!$flow->hasStore(TagAware::TAG_ID) || empty($tags)) {
            return;
        }

        // get tag id
        $tagId = $flow->getStore(TagAware::TAG_ID);

        // get tag
        $tag = $flow->getData(TagAware::TAG);

        $tagData = [];
        foreach ($tags as $tag) {
            $tagData[] = [
                'id' => Uuid::randomHex(),
                'name' => $tag,
            ];
        }

        // simply create tags
        $this->tagRepository->create($tagData, $baseEvent->getContext());
    }
}
```

{% endcode %}

As you can see, several methods are already implemented:

- `__constructor`: This only defines the default expected value. This is overwritten at runtime with the actual value, that the shop owner set in the Administration.
- `getName`: Returns a unique technical name for your action.

> - `requirements`: This defines which interfaces that the action belongs to.

Reviewer: ```suggestion
- `requirements`: This defines which interfaces the action belongs to.
```

- `handleFlow`: Use this method to handle your action stuff.
    - Use `$flow->getStore($key)` if you want to get the data from aware interfaces. E.g: `tag_id` in `TagAware`, `customer_id` from `CustomerAware` and so on.
    - Use `$flow->getData($key)` if you want to get the data from original events or additional data. E.g: `tag`, `customer`, `contactFormData` and so on.

And we also need to register this action in the container as a service, make sure you have defined a tag `<tag name="flow.action" priority="600">` at `<plugin root>/src/Resources/config/services.xml`, that your action would be added to response of *`/api/_info/flow-actions.json`* API and `priority` will decide the order of action of API response:

{% hint style="info" %}
  Available starting with Shopware 6.4.6.0

> {% endhint %}

Reviewer: ```suggestion
{% endhint %}

```

</service>
```

{% hint style="info" %}
  Available starting with Shopware 6.5.0.0

> {% endhint %}

Reviewer: ```suggestion
{% endhint %}

```
